big-number/src/main/java/BigNumber.java

```diff
@@ -0,0 +1,230 @@
+public class BigNumber {
+
+    private static final BigNumber BIG_NUMBER_ZERO = new BigNumber("0", true);
+
+    private String absValue;
+    private boolean positive;
+
+    // public constructor with all checks
+    public BigNumber(String num) {
+
+        if (num == null) {
+            throw new NumberFormatException("String must not be NULL!");
+        }
+
+        num.trim();
+        if (num.length() == 0) {
+            throw new NumberFormatException("String must not be empty!");
+        }
+
+        positive = (num.charAt(0) == '-') ? false : true;
+
+        int i = positive ? 0 : 1;
+        boolean startOfLine = true;
+
+        for (int j = i; j < num.length(); j++) {
+            int digit = charToDigit(num.charAt(j));
+            if (digit < 0 || digit > 9) {
+                throw new NumberFormatException(num + " is not a number");
+            } else if (digit == 0 && startOfLine) {
+                i++;
+            } else {
+                startOfLine = false;
+            }
+        }
+
+        // case   "00" = "0",       case   "012" = "12"
+        // case  "-00" - INVALID,   case  "-012" = "-12"
+        if (!positive && num.length() == i) {
+            throw new NumberFormatException(num + " is not a number");
+        } else {
+            absValue = (i < num.length()) ? num.substring(i) : "0";
+        }
+    }
+
+    // private constructor for internal use
+    private BigNumber(String validAbsValue, boolean isPositive) {
+        absValue = validAbsValue;
+        positive = isPositive;
+    }
+
+    private int charToDigit(char ch) {
+        return ch - '0';
+    }
+
+    // "123" -> "000123"
+    private String addSomeZeros(String str, int length) {
```

Comment:
This is internal method, you could return StringBuilder if you know you gonna need it further. Or you could pass `StringBuilder` instead of `String` to make it clear.